parser/rootfs_parser.go

```diff
@@ -0,0 +1,346 @@
+	for _, u := range upd {
+		log.Infof("calculating checksum for: %v", u.Name())
+		sum, err := calcChecksum(filepath.Join(src, u.Name()))
+		if err != nil {
+			return err
+		}
+		log.Infof("checksum for: %v [%v]", u.Name(), string(sum))
+		a := archiver.NewStreamArchiver(sum, filepath.Join(dir, withoutExt(u.Name())+".sha256sum"))
+		if err := a.Archive(tw); err != nil {
+			return errors.Wrapf(err, "reader: error storing checksum")
+		}
+	}
+	return nil
+}
+
+func (rp *RootfsParser) ArchiveHeader(tw *tar.Writer, srcDir, dstDir string) error {
+	log.Infof("processing header: %v [%v]", srcDir, dstDir)
+	if err := hFormatPreWrite.CheckHeaderStructure(srcDir); err != nil {
+		return err
+	}
+
+	// here we should get list of all update files which are
+	// part of current update
+	updFiles, err := ioutil.ReadDir(filepath.Join(srcDir, "data"))
+	if err != nil {
+		return err
+	}
+
+	// store data files
+	for _, f := range updFiles {
+		rp.updates[withoutExt(f.Name())] =
+			rootfsFile{
+				name: f.Name(),
+				path: filepath.Join(srcDir, "data", f.Name()),
+			}
+	}
+
+	log.Infof("update files: %+v", updFiles)
+
+	//TODO: use stored data
+	if err = archiveFiles(tw, updFiles, dstDir); err != nil {
+		return errors.Wrapf(err, "parser: can not store files")
+	}
+
+	a := archiver.NewFileArchiver(filepath.Join(srcDir, "type-info"),
+		filepath.Join(dstDir, "type-info"))
+	if err := a.Archive(tw); err != nil {
+		return errors.Wrapf(err, "parser: can not store type-info")
+	}
+
+	a = archiver.NewFileArchiver(filepath.Join(srcDir, "meta-data"),
+		filepath.Join(dstDir, "meta-data"))
+	if err := a.Archive(tw); err != nil {
+		return errors.Wrapf(err, "parser: can not store meta-data")
+	}
+
+	if err := archiveChecksums(tw, updFiles,
+		filepath.Join(srcDir, "data"),
+		filepath.Join(dstDir, "checksums")); err != nil {
+		return err
+	}
+
+	//TODO: get rid of bad Joins
+	for _, u := range updFiles {
+		a = archiver.NewFileArchiver(filepath.Join(srcDir, "signatures", withoutExt(u.Name())+".sig"),
+			filepath.Join(dstDir, "signatures", withoutExt(u.Name())+".sig"))
+		if err := a.Archive(tw); err != nil {
+			return errors.Wrapf(err, "parser: can not store signatures")
+		}
+	}
+
+	//TODO: scripts
+
+	return nil
+}
+
+func (rp *RootfsParser) ParseHeader(tr *tar.Reader, hPath string) error {
+	// iterate through tar archive untill some error occurs or we will
+	log.Info("processing rootfs image header")
+
+	if tr == nil {
+		return errors.New("rootfs updater: uninitialized tar reader")
+	}
+	// reach end of archive
+	for i := 0; ; i++ {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			// we have reached end of archive
+			log.Debug("rootfs updater: reached end of archive")
+			return nil
+		}
+
+		relPath, err := filepath.Rel(hPath, hdr.Name)
+		if err != nil {
+			return err
+		}
+
+		log.Infof("processing rootfs image header file: %v %v", hdr.Name, hdr.Linkname)
+
+		switch {
+		case i == 0 && strings.Compare(relPath, "files") == 0:
```

Comment:
may be a good idea to put fixed indexes in `const` (even a local one) to label what they mean